src/HaskellerAnswers.hs

```diff
@@ -43,14 +54,62 @@ data Event
     deriving stock (Eq, Show)
 
 updateApp :: Event -> Model -> Effect Event Model
-updateApp event model = case event of
+updateApp event model@Model{..} = case event of
     NoEvent    -> noEff model
-    NextAnswer -> let cur :| rest = modelAllAnswers model in
-        noEff $ model
-            { modelAllAnswers    = NE.fromList rest
-            , modelCurrentAnswer = cur
+    NextAnswer ->
+        let (cur, newGiven, newRemaining, newGen) = extractWithBackup
+                modelGivenAnswers
+                modelRemainingAnswers
+                modelRandGen
+        in noEff $ model
+            { modelRemainingAnswers = newRemaining
+            , modelGivenAnswers     = newGiven
+            , modelCurrentAnswer    = cur
+            , modelRandGen          = newGen
             }
 
+{- | This function extracts random element from a given list and moves it to
+another list (backup). If the given list where we want to exract is empty, then
+we extract from backup and empty our backup.
+
+This function returns:
+
+1. Extracted random element
+2. A possibly new backup (@backup@ is not changed if @l@ is not empty)
+3. A list of answers without extracted element
+4. A new random generator
+
+__Invariant__: this function hangs if both lists are empty. It assumes that at
+least one list is not empty.
+-}
+extractWithBackup :: [a] -> [a] -> StdGen -> (a, [a], [a], StdGen)
+extractWithBackup backup l gen =
+    let (pick, newGen) = extractRandom gen l
+    in case pick of
+        Nothing      -> extractWithBackup [] backup newGen
+        Just (x, xs) -> (x, x:backup, xs, newGen)
```

Comment:
@chshersh is it necessary to keep the backup list? We only use it when the current list is empty and it always equals to the default list (not by order though). So I was thinking probably we can omit tracking that?

Reply:
@vrom911 Nice catch!!! I missed this thing. Yeah, we have a global immutable list of all default answers. This will simplify things a lot!!!